src/ophys_etl/modules/decrosstalk/qc_plotting.py

```diff
@@ -0,0 +1,1147 @@
+
+    fig: matplotlib.figure.Figure
+        The figure in which we are plotting
+
+    Returns
+    -------
+    None
+    """
+
+    green = matplotlib.colors.to_rgba('green')
+    green = tuple([int(255*g) for g in green[:3]])
+
+    red = matplotlib.colors.to_rgba('red')
+    red = tuple([int(255*r) for r in red[:3]])
+
+    outlier_color = (255, 108, 220)
+    valid_color = (255, 196, 92)
+
+    outlier_color_hex = '#%02x%02x%02x' % outlier_color
+    valid_color_hex = '#%02x%02x%02x' % valid_color
+    green_hex = '#%02x%02x%02x' % green
+    red_hex = '#%02x%02x%02x' % red
+
+    inner_grid = gridspec.GridSpecFromSubplotSpec(2, 4,
+                                  subplot_spec=subplot_spec,    # noqa E128
+                                  wspace=0.05, hspace=0.)       # noqa E128
+
+    # loop over the two planes
+    for ii in range(2):
+        plane = ophys_planes[ii]
+
+        with h5py.File(plane.qc_file_path, 'r') as in_file:
+            if 'ROI' not in in_file:
+                # There were no ROIs in this plane; just move on
+                continue
+            roi_keys = list(in_file['ROI'].keys())
+
+        roi_id = np.array([int(k) for k in roi_keys])
+        roi_min = roi_id.min()
+
+        avg_mixing_matrix = get_avg_mixing_matrix(plane)
+        roi_qc = find_problematic_rois(plane, roi_flags)
+        n_valid_roi = 0
+        n_ghost_roi = 0
+
+        # load the raw max projection image
+        raw_img = PIL.Image.open(plane.maximum_projection_image_path)
+        n_rows = raw_img.size[0]
+        n_cols = raw_img.size[1]
+        raw_img = np.array(raw_img).reshape(n_rows, n_cols)
+
+        # convert to an RGBA image
+        max_img = np.zeros((n_rows, n_cols, 4), dtype=int)
+
+        for jj in range(3):
+            max_img[:, :, jj] = raw_img[:, :]
+        max_img[:, :, 3] = 255
+
+        # superimpose gridlines
+        max_img = add_gridlines(max_img, 3)
+
+        axes = []
+        for jj in range(3):
+            ax = plt.Subplot(fig, inner_grid[ii, jj])
+            ax.set_xlim(0, 35+n_cols)
+            ax.set_ylim(0, 35+n_rows)
+            ax.patch.set_alpha(0)
+            for s in ('top', 'bottom', 'left', 'right'):
+                ax.spines[s].set_visible(False)
+                ax.tick_params(which='both', axis='both',
+                               left=0, bottom=0,
+                               labelleft=0, labelbottom=0)
+            axes.append(ax)
+
+        extent = (20, 20+n_cols,
+                  20, 20+n_rows)
+
+        # create three separate copies of the max_img
+        # so that we only have to loop over the ROIs once
+        max_img_copies = {}
+        max_img_copies[0] = np.copy(max_img)
+        max_img_copies[2] = np.copy(max_img)
+
+        axes[1].imshow(np.copy(max_img),
+                       extent=extent)
+
+        # loop over ROIS, adding centroids and masks to the copied
+        # maximum projection images
+        for roi in plane.roi_list:
```

Comment:
this is a pretty long and nested mix of collecting data and modifying plots. Is there a way to separate these concerns, perhaps in small functions, to make it quicker to understand and maintain later?

Reply:
I've tried to break it down a little more. Specifically, I've added/factored out

`find_roi_colors` -- which moves all of the code assigning colors to ROIs into the method that formerly just flitered ROIs by residual crosstalk

`get_max_projection_image` -- which loads, formats, and adds the gridlines to the full max projection image for an experimental plane

`plot_roi_mask` -- which handles the plotting of an ROI onto the max projection image of its experimental plane

`plot_summary_statistics` -- which plots the final panel in the summary plot (the one that just lists the number of ghosts along with the average mixing matrix for the experimental plane)

I also factored out the part of `plot_pair_of_rois` that created thumbnails from the max projection image into `get_img_thumbnails`. I don't think there is much more that can be factored out of `plot_pair_of_rois`.

I'm a little worried that the methods supporting the two different QC plots are starting to sound very similar, even though they do slightly different things to the point where I don't think I can reuse the code very efficiently. I am tempted to create two separate modules, one for each type of QC figure, with a `qc_plotting_utils.py` for the handful of shared functions between the two.

Does that seem reasonable to you, or would it be overkill?

Reply:
I think all of those ideas sound reasonable -
* shared code where possible
* separate modules for separate artifacts (as long as the duplicated data loading isn't too onerous)

Reply:
I have created a subdirectory `qc_plotting` under `decrosstalk`. It contains three files.

`utils.py` contains the one method both other files need
`summary_plot.py` contains the code to generate the figure displaying all eight planes at once
`pairwise_plot.py` contains the code to generate the figures comparing pairs of overlapping ROIs

The user-facing methods can be imported as
```
from ophys_etl.modules.decrosstalk.qc_plotting import generate_roi_figure, generate_pairwise_figures
```